File: app/src/main/java/knf/kuma/commons/BypassUtil.java
package knf.kuma.commons;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;
import android.webkit.CookieManager;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import org.jsoup.Connection;
import org.jsoup.HttpStatusException;
import org.jsoup.Jsoup;

import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import xdroid.toaster.Toaster;

/**
 * Created by jordy on 17/03/2018.
 */

public class BypassUtil extends AppCompatActivity {
    public static final String userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:57.0) Gecko/20100101 Firefox/57.0";
    public static boolean isLoading = false;

    @BindView(R.id.webview)
    WebView web;

    public static void check(final Activity activity) {
        start(activity);
        /*AsyncTask.execute(() -> {
            if (isNeeded(activity) && !isLoading) {
                isLoading = true;
                activity.runOnUiThread(() -> {
                    Log.e("CloudflareBypass", "is needed");
                    clearCookies();
                    final WebView webView = new WebView(activity);
                    webView.getSettings().setJavaScriptEnabled(true);
                    webView.setWebViewClient(new WebViewClient() {
                        @Override
                        public boolean shouldOverrideUrlLoading(final WebView view, final String url) {
                            Log.e("CloudflareBypass", "Override " + url);
                            if (url.equals("https://animeflv.net/")) {
                                Log.e("CloudflareBypass", "Cookies: " + CookieManager.getInstance().getCookie("https://animeflv.net/"));
                                saveCookies(activity);
                                Toaster.toast("Bypass actualizado");
                                PicassoSingle.clear();
                                ((BypassListener) activity).onNeedRecreate();
                            }
                            isLoading = false;
                            return false;
                        }

                        @Override
                        public void onPageStarted(WebView view, String url, Bitmap favicon) {
                            super.onPageStarted(view, url, favicon);
                            Log.e("CloudflareBypass", "Started " + url);
                        }
                    });
                    webView.getSettings().setUserAgentString(userAgent);
                    webView.loadUrl("https://animeflv.net/");
                });
            } else Log.e("CloudflareBypass", "Not needed");
        });*/
    }

    public static void saveCookies(Context context) {
        String cookies = CookieManager.getInstance().getCookie("https://animeflv.net/").trim();
        if (cookies.contains("cf_clearance")) {
            String[] parts = cookies.split(";");
            SharedPreferences.Editor preferences = PreferenceManager.getDefaultSharedPreferences(context).edit();
            for (String cookie : parts) {
                if (cookie.contains("__cfduid"))
                    preferences.putString("__cfduid", cookie.trim().substring(cookie.trim().indexOf("=") + 1));
                if (cookie.contains("cf_clearance"))
                    preferences.putString("cf_clearance", cookie.trim().substring(cookie.trim().indexOf("=") + 1));
            }
            preferences.apply();
        }
    }

    public static void clearCookies() {
        CookieManager cookieManager = CookieManager.getInstance();
        String cookiestring = cookieManager.getCookie(".animeflv.net");
        if (cookiestring != null) {
            String[] cookies = cookiestring.split(";");
            for (String cookie : cookies) {
                String[] cookieparts = cookie.split("=");
                cookieManager.setCookie(".animeflv.net", cookieparts[0].trim() + "=; Expires=Wed, 31 Dec 2025 23:59:59 GMT");
            }
        }
    }

    public static boolean isNeeded(Context context) {
        try {
            Connection.Response response = Jsoup.connect("https://animeflv.net/").cookies(getMapCookie(context)).userAgent(BypassUtil.userAgent).execute();
            return response.statusCode() == 503;
        } catch (HttpStatusException e) {
            return e.getStatusCode() == 503;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
    }

    public static Map<String, String> getMapCookie(Context context) {
        Map<String, String> map = new LinkedHashMap<>();
        map.put("device", "computer");
        map.put("cf_clearance", getClearance(context));
        map.put("__cfduid", getCFDuid(context));
        return map;
    }

    public static String getStringCookie(Context context) {
        return "device=computer; " +
                "cf_clearance=" + getClearance(context) + "; " +
                "__cfduid=" + getCFDuid(context);
    }

    private static String getClearance(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context).getString("cf_clearance", "00000000");
    }

    private static String getCFDuid(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context).getString("__cfduid", "00000000");
    }

    public static void start(Context context) {
        context.startActivity(new Intent(context, BypassUtil.class));
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_web);
        ButterKnife.bind(this);
        AsyncTask.execute(() -> {
            if (isNeeded(this) && !isLoading) {
                Toaster.toast("Creando bypass");
                isLoading = true;
                Log.e("CloudflareBypass", "is needed");
                clearCookies();
                web.getSettings().setJavaScriptEnabled(true);
                web.setWebViewClient(new WebViewClient() {
                    @Override
                    public boolean shouldOverrideUrlLoading(final WebView view, final String url) {
                        Log.e("CloudflareBypass", "Override " + url);
                        if (url.equals("https://animeflv.net/")) {
                            Log.e("CloudflareBypass", "Cookies: " + CookieManager.getInstance().getCookie("https://animeflv.net/"));
                            saveCookies(BypassUtil.this);
                            Toaster.toast("Bypass actualizado");
                            PicassoSingle.clear();
                        }
                        isLoading = false;
                        Toaster.toast("Bypass creado");
                        finish();
                        return false;
                    }
                });
                web.getSettings().setUserAgentString(userAgent);
                web.loadUrl("https://animeflv.net/");
            } else {
                Log.e("CloudflareBypass", "Not needed");
                Toaster.toast("Bypass no necesario");
                finish();
            }
        });
    }

    public interface BypassListener {
        void onNeedRecreate();
    }
}

File: app/src/main/java/knf/kuma/commons/BypassUtil.kt
package knf.kuma.commons

import android.content.Context
import android.webkit.CookieManager
import android.webkit.WebSettings
import android.webkit.WebView
import androidx.preference.PreferenceManager
import knf.kuma.App
import knf.kuma.ads.AdsUtils
import knf.kuma.uagen.UAGenerator
import knf.kuma.uagen.randomUA
import knf.tools.bypass.DisplayType
import knf.tools.bypass.Request
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jsoup.HttpStatusException
import org.jsoup.Jsoup

/**
 * Created by jordy on 17/03/2018.
 */

class BypassUtil {

    interface BypassListener {
        fun onNeedRecreate()
    }

    companion object {
        val userAgent: String
            get() = if (PrefsUtil.useDefaultUserAgent && !PrefsUtil.alwaysGenerateUA) noCrashLet { WebSettings.getDefaultUserAgent(App.context) }
                ?: PrefsUtil.userAgent else PrefsUtil.userAgent
        var isLoading = false
        var isChecking = false

        private const val keyCfClearance = "cf_clearance"
        private const val keyCfDuid = "__cfduid"
        private const val keyCookiesBypass = "bypass_cookies"
        private const val defaultValue = ""
        const val testLink = "https://www3.animeflv.net/"

        fun createRequest(): Request {
            return Request(
                testLink,
                lastUA = PrefsUtil.userAgent,
                showReload = AdsUtils.remoteConfigs.getBoolean("bypass_show_reload"),
                useFocus = isTV,
                maxTryCount = AdsUtils.remoteConfigs.getLong("bypass_max_tries").toInt(),
                useLatestUA = true,
                reloadOnCaptcha = false,
                waitCaptcha = true,
                clearCookiesAtStart = true,
                displayType = DisplayType.DIALOG,
                dialogStyle = 0
            )
        }

        suspend fun clearCookiesIfNeeded() {
            if (!withContext(Dispatchers.IO) { isCloudflareActive() })
                clearCookies(null)
        }

        fun saveCookies(context: Context, cookies: String): Boolean =
            noCrashLet(false) {
                bypassCookies = cookies
                if (cookies.contains(keyCfClearance)) {
                        val parts = cookies.split(";").dropLastWhile { it.isEmpty() }.toTypedArray()
                        for (cookie in parts) {
                            if (cookie.contains(keyCfDuid))
                                setCFDuid(context, cookie.trim().substring(cookie.trim().indexOf("=") + 1))
                            if (cookie.contains(keyCfClearance)) {
                                val clearance = cookie.trim().substring(cookie.trim().indexOf("=") + 1)
                                if (clearance.isBlank())
                                    return@noCrashLet false
                                setClearance(context, clearance)
                            }
                        }
                        return@noCrashLet true
                    }
                    false
                }

        fun clearCookies(webView: WebView?) {
            noCrash {
                val cookieManager = CookieManager.getInstance()
                cookieManager.removeAllCookies(null)
                webView?.clearCache(true)
                bypassCookies = null
                setCFDuid(App.context)
                setClearance(App.context)
                PrefsUtil.userAgent = UAGenerator.getRandomUserAgent()
            }
        }

        fun isNeeded(url: String = testLink): Boolean {
            return try {
                val response = jsoupCookies(url).execute()
                response.statusCode().let { it == 503 || it == 403 }
            } catch (e: HttpStatusException) {
                e.statusCode.let { it == 503 || it == 403 }
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

        fun isCloudflareActive(url: String = testLink): Boolean {
            return try {
                val response = Jsoup.connect(url).followRedirects(true).execute()
                response.statusCode().let { it == 503 || it == 403 }
            } catch (e: HttpStatusException) {
                e.statusCode.let { it == 503 || it == 403 }
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

        fun isCloudflareActiveRandom(url: String = testLink): Boolean {
            return try {
                val response = Jsoup.connect(url).followRedirects(true).userAgent(randomUA()).execute()
                response.statusCode().let { it == 503 || it == 403 }
            } catch (e: HttpStatusException) {
                e.statusCode.let { it == 503 || it == 403 }
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

        fun isNeededFlag(): Int {
            return try {
                val response = jsoupCookies(testLink).execute()
                when (response.statusCode()) {
                    503 -> 1
                    403 -> 2
                    else -> 0
                }
            } catch (e: HttpStatusException) {
                when (e.statusCode) {
                    503 -> 1
                    403 -> 2
                    else -> 0
                }
            } catch (e: Exception) {
                e.printStackTrace()
                0
            }
        }

        fun getMapCookie(context: Context): Map<String, String> {
            val map = LinkedHashMap<String, String>()
            map["device"] = "computer"
            map["InstiSession"] =
                "eyJpZCI6IjRlNGYwNWYxLTg4NDMtNGQwOS05ODlmLWM1OWQ5N2NmNjVlYyIsInJlZmVycmVyIjoiIiwiY2FtcGFpZ24iOnsic291cmNlIjpudWxsLCJtZWRpdW0iOm51bGwsImNhbXBhaWduIjpudWxsLCJ0ZXJtIjpudWxsLCJjb250ZW50IjpudWxsfX0="
            bypassCookies?.split(";")?.forEach {
                if (it.contains("=")) {
                    val split = it.split("=")
                    map[split[0]] = split[1]
                }
            }
            //getClearance(context).let { if (it != defaultValue) map[keyCfClearance] = it }
            //getCFDuid(context).let { if (it != defaultValue) map[keyCfDuid] = it }
            return map
        }

        fun getStringCookie(context: Context): String {
            val builder = StringBuilder()
            for ((key, value) in getMapCookie(context))
                builder.append("$key=$value;")
            return builder.toString().dropLastWhile { it == ' ' || it == ';' }
        }

        var bypassCookies: String?
            set(value) = PreferenceManager.getDefaultSharedPreferences(App.context).edit().putString(keyCookiesBypass, value).apply()
            get() = PreferenceManager.getDefaultSharedPreferences(App.context).getString(keyCookiesBypass, null)

        fun getClearance(context: Context): String {
            return PreferenceManager.getDefaultSharedPreferences(context).getString(keyCfClearance, defaultValue)
                ?: defaultValue
        }

        private fun setClearance(context: Context, value: String = defaultValue) {
            PreferenceManager.getDefaultSharedPreferences(context).edit().putString(keyCfClearance, value).apply()
        }

        fun getCFDuid(context: Context): String {
            return PreferenceManager.getDefaultSharedPreferences(context).getString(keyCfDuid, defaultValue)
                    ?: defaultValue
        }

        private fun setCFDuid(context: Context, value: String = defaultValue) {
            PreferenceManager.getDefaultSharedPreferences(context).edit().putString(keyCfDuid, value).apply()
        }
    }
}
